File: Maze Generator - Recursive Back tracker (C#, Visual Basic)/[C#]-Maze Generator - Recursive Back tracker (C#, Visual Basic)/C#/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using System.Collections;
using System.Diagnostics;
namespace WindowsFormsApplication1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            doLayout();
        }
        public void doLayout()
        {
            Panel2.Top = 100;
            Panel2.Left = 0;
            Panel2.Height = this.ClientRectangle.Height - Panel2.Top;
            Panel2.Width = this.ClientRectangle.Width;
            Panel2.BorderStyle = BorderStyle.FixedSingle;
        }
        private void Button1_Click(object sender, EventArgs e)
        {
            Maze maze = new Maze(Convert.ToInt32(nudRows.Value), Convert.ToInt32(nudCols.Value), Convert.ToInt32(nudWidth.Value), Convert.ToInt32(nudHeight.Value));
            maze.MazeComplete += (Image m) =>{
                Panel1.BackgroundImage = m;
                Panel1.BackgroundImageLayout = ImageLayout.None;
                Panel1.Width = m.Width;
                Panel1.Height = m.Height;
                //maze.PrintMaze()
            };
            maze.Generate();
        }

        private void Form1_Resize(object sender, EventArgs e)
        {
            doLayout();
        }
    }
}

public class Maze : Control
{
    int Rows;
    int Columns;
    int cellWidth;
    int cellHeight;
    Dictionary<string, Cell> cells = new Dictionary<string, Cell>();
    Stack<Cell> stack = new Stack<Cell>();

    public Image MazeImage;
    public event MazeCompleteEventHandler MazeComplete;
    public delegate void MazeCompleteEventHandler(Image Maze);
    private event CallCompleteEventHandler CallComplete;
    private delegate void CallCompleteEventHandler(Image Maze);
    public new Rectangle Bounds
    {
        get
        {
            Rectangle rect = new Rectangle(0, 0, Width, Height);
            return rect;
        }
    }
    private System.Drawing.Printing.PrintDocument withEventsField_printDoc = new System.Drawing.Printing.PrintDocument();
    public System.Drawing.Printing.PrintDocument printDoc
    {
        get { return withEventsField_printDoc; }
        set
        {
            if (withEventsField_printDoc != null)
            {
                withEventsField_printDoc.PrintPage -= PrintImage;
            }
            withEventsField_printDoc = value;
            if (withEventsField_printDoc != null)
            {
                withEventsField_printDoc.PrintPage += PrintImage;
            }
        }
    }
    private void PrintImage(object sender, System.Drawing.Printing.PrintPageEventArgs e)
    {
        List<string> nonprinters = new string[] { "Send To OneNote 2013", "PDFCreator", "PDF Architect 4", "Microsoft XPS Document Writer", "Microsoft Print to PDF", "Fax", "-" }.ToList();
        string printerName = "none";
        foreach (string a in System.Drawing.Printing.PrinterSettings.InstalledPrinters)
        {
            if (nonprinters.IndexOf(a) > -1)
                continue;
            printerName = a;
        }
        if (printerName == "none")
            return;
        printDoc.PrinterSettings.PrinterName = printerName;
        int imageLeft = Convert.ToInt32(e.PageBounds.Width / 2) - Convert.ToInt32(MazeImage.Width / 2);
        int imageTop = Convert.ToInt32(e.PageBounds.Height / 2) - Convert.ToInt32(MazeImage.Height / 2);
        e.Graphics.DrawImage(MazeImage, imageLeft, imageTop);
    }
    public void PrintMaze()
    {
        printDoc.Print();
    }
    public void Generate()
    {
        int c = 0;
        int r = 0;
        for (int y = 0; y <= Height; y += cellHeight)
        {
            for (int x = 0; x <= Width; x += cellWidth)
            {
                Cell cell = new Cell(new Point(x, y), new Size(cellWidth, cellHeight),ref cells, r, c, (Rows - 1), (Columns - 1));
                c += 1;
            }
            c = 0;
            r += 1;
        }
        System.Threading.Thread thread = new System.Threading.Thread(Dig);
        thread.Start();
    }
    private void Dig()
    {
        int r = 0;
        int c = 0;
        string key = "c" + 5 + "r" + 5;
        Cell startCell = cells[key];
        stack.Clear();
        startCell.Visited = true;
        while ((startCell != null))
        {
            startCell = startCell.Dig(ref stack);
            if (startCell != null)
            {
                startCell.Visited = true;
                startCell.Pen = Pens.Black;
            }
        }
        stack.Clear();
        Bitmap Maze = new Bitmap(Width, Height);
        using (Graphics g = Graphics.FromImage(Maze))
        {
            g.Clear(Color.White);
            if (cells.Count > 0)
            {
                for (r = 0; r <= this.Rows - 1; r++)
                {
                    for (c = 0; c <= this.Columns - 1; c++)
                    {
                        Cell cell = cells["c" + c + "r" + r];
                        cell.draw(g);
                    }
                }
            }
        }
        this.MazeImage = Maze;

        if (CallComplete != null)
        {
            CallComplete(Maze);
        }
    }
    public delegate void dComplete(Image maze);
    private void Call_Complete(Image maze)
    {
        if (this.InvokeRequired)
        {
            this.Invoke(new dComplete(Call_Complete), maze);
        }
        else
        {
            if (MazeComplete != null)
            {
                MazeComplete(maze);
            }
        }
    }
    public Maze(int rows, int columns, int cellWidth, int cellHeight)
    {
        CallComplete += Call_Complete;
        this.Rows = rows;
        this.Columns = columns;
        this.cellWidth = cellWidth;
        this.cellHeight = cellHeight;
        this.Width = (this.Columns * this.cellWidth) + 1;
        this.Height = (this.Rows * this.cellHeight) + 1;
        this.CreateHandle();
    }
}
public class Cell
{
    public bool NorthWall = true;
    public bool SouthWall = true;
    public bool WestWall = true;
    public bool EastWall = true;
    public string id;
    public Pen Pen = Pens.Black;
    public Rectangle Bounds;
    public Dictionary<string, Cell> Cells;
    public int Column;
    public int Row;
    public string NeighborNorthID;
    public string NeighborSouthID;
    public string NeighborEastID;
    public string NeighborWestID;
    public bool Visited = false;
    public Stack<Cell> Stack;
    public void draw(Graphics g)
    {
        if (NorthWall) g.DrawLine(Pen, new Point(Bounds.Left, Bounds.Top), new Point(Bounds.Right, Bounds.Top));
        if (SouthWall)g.DrawLine(Pen, new Point(Bounds.Left, Bounds.Bottom), new Point(Bounds.Right, Bounds.Bottom));
        if (WestWall)g.DrawLine(Pen, new Point(Bounds.Left, Bounds.Top), new Point(Bounds.Left, Bounds.Bottom));
        if (EastWall)g.DrawLine(Pen, new Point(Bounds.Right, Bounds.Top), new Point(Bounds.Right, Bounds.Bottom));
    }
    public Cell(Point location, Size size, ref Dictionary<string, Cell> cellList, int r, int c, int maxR, int maxC)
    {
        this.Bounds = new Rectangle(location, size);
        this.Column = c;
        this.Row = r;
        this.id = "c" + c + "r" + r;
        int rowNort = r - 1;
        int rowSout = r + 1;
        int colEast = c + 1;
        int colWest = c - 1;
        NeighborNorthID = "c" + c + "r" + rowNort;
        NeighborSouthID = "c" + c + "r" + rowSout;
        NeighborEastID = "c" + colEast + "r" + r;
        NeighborWestID = "c" + colWest + "r" + r;
        if (rowNort < 0) NeighborNorthID = "none";
        if (rowSout > maxR) NeighborSouthID = "none";
        if (colEast > maxC) NeighborEastID = "none";
        if (colWest < 0) NeighborWestID = "none";
        this.Cells = cellList;
        this.Cells.Add(this.id, this);
    }
    public Cell getNeighbor()
    {
        List<Cell> c = new List<Cell>();
        if (!(NeighborNorthID == "none") && Cells[NeighborNorthID].Visited == false)c.Add(Cells[NeighborNorthID]);
        if (!(NeighborSouthID == "none") && Cells[NeighborSouthID].Visited == false)c.Add(Cells[NeighborSouthID]);
        if (!(NeighborEastID == "none") && Cells[NeighborEastID].Visited == false)c.Add(Cells[NeighborEastID]);
        if (!(NeighborWestID == "none") && Cells[NeighborWestID].Visited == false)c.Add(Cells[NeighborWestID]);
        int max = c.Count;
        Cell currentCell = null;
        if (c.Count > 0)
        {
            Microsoft.VisualBasic.VBMath.Randomize();
            int index = Convert.ToInt32(Conversion.Int(c.Count * VBMath.Rnd()));
            currentCell = c[index];
        }
        return currentCell;
    }
    public Cell Dig(ref Stack<Cell> stack)
    {
        this.Stack = stack;
        Cell nextCell = getNeighbor();
        if ((nextCell != null))
        {
            stack.Push(nextCell);
            if (nextCell.id == this.NeighborNorthID)
            {
                this.NorthWall = false;
                nextCell.SouthWall = false;
            }
            else if (nextCell.id == this.NeighborSouthID)
            {
                this.SouthWall = false;
                nextCell.NorthWall = false;
            }
            else if (nextCell.id == this.NeighborEastID)
            {
                this.EastWall = false;
                nextCell.WestWall = false;
            }
            else if (nextCell.id == this.NeighborWestID)
            {
                this.WestWall = false;
                nextCell.EastWall = false;
            }
        }
        else if (!(stack.Count == 0))
        {
            nextCell = stack.Pop();
        }
        else
        {
            return null;
        }
        return nextCell;
    }
}

File: Maze Generator - Recursive Back tracker (C#, Visual Basic)/[VB]-Maze Generator - Recursive Back tracker (C#, Visual Basic)/VB/Form1.vb
Option Strict On
Option Explicit On
Option Infer Off
Public Class Form1
    Sub doLayout()
        Panel2.Top = 100
        Panel2.Left = 0
        Panel2.Height = Me.ClientRectangle.Height - Panel2.Top
        Panel2.Width = Me.ClientRectangle.Width
        Panel2.BorderStyle = BorderStyle.FixedSingle
    End Sub
    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        doLayout()
    End Sub
    Private Sub Form1_Resize(sender As Object, e As EventArgs) Handles Me.Resize
        doLayout()
    End Sub
    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        Dim maze As New Maze(CInt(nudRows.Value), CInt(nudCols.Value), CInt(nudWidth.Value), CInt(nudHeight.Value))
        AddHandler maze.MazeComplete, Sub(m As Image)
                                          Panel1.BackgroundImage = m
                                          Panel1.BackgroundImageLayout = ImageLayout.None
                                          Panel1.Width = m.Width
                                          Panel1.Height = m.Height
                                          'maze.PrintMaze()
                                      End Sub
        maze.Generate()

    End Sub
End Class
Public Class Maze
    Inherits Control
    Dim Rows As Integer
    Dim Columns As Integer
    Dim cellWidth As Integer
    Dim cellHeight As Integer
    Dim cells As New Dictionary(Of String, Cell)
    Dim stack As New Stack(Of Cell)
    Public Maze As Image

    Public Event MazeComplete(Maze As Image)
    Private Event CallComplete(Maze As Image)
    Public Shadows ReadOnly Property Bounds As Rectangle
        Get
            Dim rect As New Rectangle(0, 0, Width, Height)
            Return rect
        End Get
    End Property
    Dim WithEvents printDoc As New Printing.PrintDocument()
    Private Sub PrintImage(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintPageEventArgs) Handles printDoc.PrintPage
        Dim nonprinters As List(Of String) = ({"Send To OneNote 2013", "PDFCreator", "PDF Architect 4",
                               "Microsoft XPS Document Writer", "Microsoft Print to PDF", "Fax", "-"}).ToList
        Dim printerName As String = "none"
        For Each a As String In System.Drawing.Printing.PrinterSettings.InstalledPrinters
            If nonprinters.IndexOf(a) > -1 Then Continue For
            printerName = a
        Next
        If printerName = "none" Then Exit Sub
        printDoc.PrinterSettings.PrinterName = printerName
        Dim imageLeft As Integer = CInt(e.PageBounds.Width / 2) - CInt(Maze.Width / 2)
        Dim imageTop As Integer = CInt(e.PageBounds.Height / 2) - CInt(Maze.Height / 2)
        e.Graphics.DrawImage(Maze, imageLeft, imageTop)
    End Sub
    Public Sub PrintMaze()
        printDoc.Print()
    End Sub
    Public Sub Generate()
        Dim c As Integer = 0
        Dim r As Integer = 0
        For y As Integer = 0 To Height Step cellHeight
            For x As Integer = 0 To Width Step cellWidth
                Dim cell As New Cell(New Point(x, y), New Size(cellWidth, cellHeight), cells, r, c, (Rows - 1), (Columns - 1))
                c += 1
            Next
            c = 0 : r += 1
        Next
        Dim thread As New Threading.Thread(AddressOf Dig)
        thread.Start()
    End Sub
    Private Sub Dig()
        Dim r As Integer = 0
        Dim c As Integer = 0
        Dim key As String = "c" & 5 & "r" & 5
        Dim startCell As Cell = cells(key)
        stack.Clear()
        startCell.Visited = True
        While (startCell IsNot Nothing)
            startCell = startCell.Dig(stack)
            If startCell IsNot Nothing Then
                startCell.Visited = True
                startCell.Pen = Pens.Black
            End If
        End While
        stack.Clear()
        Dim Maze As New Bitmap(Width, Height)
        Using g As Graphics = Graphics.FromImage(Maze)
            g.Clear(Color.White)
            If cells.Count > 0 Then
                For r = 0 To Me.Rows - 1
                    For c = 0 To Me.Columns - 1
                        Dim cell As Cell = cells("c" & c & "r" & r)
                        cell.draw(g)
                    Next
                Next
            End If
        End Using
        Me.Maze = Maze
        RaiseEvent CallComplete(Maze)
    End Sub
    Delegate Sub dComplete(maze As Image)
    Private Sub Call_Complete(maze As Image) Handles Me.CallComplete
        If Me.InvokeRequired Then
            Me.Invoke(New dComplete(AddressOf Call_Complete), maze)
        Else
            RaiseEvent MazeComplete(maze)
        End If
    End Sub
    Sub New(rows As Integer, columns As Integer, cellWidth As Integer, cellHeight As Integer)
        Me.Rows = rows
        Me.Columns = columns
        Me.cellWidth = cellWidth
        Me.cellHeight = cellHeight
        Me.Width = (Me.Columns * Me.cellWidth) + 1
        Me.Height = (Me.Rows * Me.cellHeight) + 1
        Me.CreateHandle()
    End Sub
End Class
Public Class Cell
    Public NorthWall As Boolean = True
    Public SouthWall As Boolean = True
    Public WestWall As Boolean = True
    Public EastWall As Boolean = True
    Public id As String '
    Public Pen As Pen = Pens.Black
    Public Bounds As Rectangle '
    Public Cells As Dictionary(Of String, Cell) '
    Public Column As Integer '
    Public Row As Integer '
    Public NeighborNorthID As String '
    Public NeighborSouthID As String '
    Public NeighborEastID As String '
    Public NeighborWestID As String '
    Public Visited As Boolean = False '
    Public Stack As Stack(Of Cell) '
    Public Sub draw(g As Graphics)
        If NorthWall Then g.DrawLine(Pen, New Point(Bounds.Left, Bounds.Top), New Point(Bounds.Right, Bounds.Top))
        If SouthWall Then g.DrawLine(Pen, New Point(Bounds.Left, Bounds.Bottom), New Point(Bounds.Right, Bounds.Bottom))
        If WestWall Then g.DrawLine(Pen, New Point(Bounds.Left, Bounds.Top), New Point(Bounds.Left, Bounds.Bottom))
        If EastWall Then g.DrawLine(Pen, New Point(Bounds.Right, Bounds.Top), New Point(Bounds.Right, Bounds.Bottom))
    End Sub
    Sub New(location As Point, size As Size, ByRef cellList As Dictionary(Of String, Cell), r As Integer, c As Integer, maxR As Integer, maxC As Integer)
        Me.Bounds = New Rectangle(location, size)
        Me.Column = c
        Me.Row = r
        Me.id = "c" & c & "r" & r
        Dim rowNort As Integer = r - 1
        Dim rowSout As Integer = r + 1
        Dim colEast As Integer = c + 1
        Dim colWest As Integer = c - 1
        NeighborNorthID = "c" & c & "r" & rowNort
        NeighborSouthID = "c" & c & "r" & rowSout
        NeighborEastID = "c" & colEast & "r" & r
        NeighborWestID = "c" & colWest & "r" & r
        If rowNort < 0 Then NeighborNorthID = "none"
        If rowSout > maxR Then NeighborSouthID = "none"
        If colEast > maxC Then NeighborEastID = "none"
        If colWest < 0 Then NeighborWestID = "none"
        Me.Cells = cellList
        Me.Cells.Add(Me.id, Me)
    End Sub
    Function getNeighbor() As Cell
        Dim c As New List(Of Cell)
        If Not NeighborNorthID = "none" AndAlso Cells(NeighborNorthID).Visited = False Then c.Add(Cells(NeighborNorthID))
        If Not NeighborSouthID = "none" AndAlso Cells(NeighborSouthID).Visited = False Then c.Add(Cells(NeighborSouthID))
        If Not NeighborEastID = "none" AndAlso Cells(NeighborEastID).Visited = False Then c.Add(Cells(NeighborEastID))
        If Not NeighborWestID = "none" AndAlso Cells(NeighborWestID).Visited = False Then c.Add(Cells(NeighborWestID))
        Dim max As Integer = c.Count
        Dim currentCell As Cell = Nothing
        If c.Count > 0 Then
            Randomize()
            Dim index As Integer = CInt(Int(c.Count * Rnd()))
            currentCell = c(index)
        End If
        Return currentCell
    End Function
    Function Dig(ByRef stack As Stack(Of Cell)) As Cell
        Me.Stack = stack
        Dim nextCell As Cell = getNeighbor()
        If Not nextCell Is Nothing Then
            stack.Push(nextCell)
            If nextCell.id = Me.NeighborNorthID Then
                Me.NorthWall = False
                nextCell.SouthWall = False
            ElseIf nextCell.id = Me.NeighborSouthID
                Me.SouthWall = False
                nextCell.NorthWall = False
            ElseIf nextCell.id = Me.NeighborEastID
                Me.EastWall = False
                nextCell.WestWall = False
            ElseIf nextCell.id = Me.NeighborWestID
                Me.WestWall = False
                nextCell.EastWall = False
            End If
        ElseIf Not stack.Count = 0
            nextCell = stack.Pop
        Else
            Return Nothing
        End If
        Return nextCell
    End Function
End Class
